packages/react/src/publication/adapters/ReactionGateway.ts

```diff
@@ -0,0 +1,94 @@
+import { ApolloClient, NormalizedCacheObject } from '@apollo/client';
+import {
+  AddReactionDocument,
+  AddReactionMutation,
+  AddReactionMutationVariables,
+  RemoveReactionDocument,
+  RemoveReactionMutation,
+  RemoveReactionMutationVariables,
+  ReactionTypes,
+} from '@lens-protocol/api-bindings';
+import { ReactionType } from '@lens-protocol/domain/entities';
+import {
+  ReactionRequest,
+  IReactionGateway,
+  NetworkError,
+} from '@lens-protocol/domain/use-cases/publications';
+import { PromiseResult, success, failure, assertError } from '@lens-protocol/shared-kernel';
+
+const reactionTypeMap: Record<ReactionType, ReactionTypes> = {
+  [ReactionType.UPVOTE]: ReactionTypes.Upvote,
+  [ReactionType.DOWNVOTE]: ReactionTypes.Downvote,
+};
+
+export class ReactionGateway implements IReactionGateway {
+  constructor(private apolloClient: ApolloClient<NormalizedCacheObject>) {}
+
+  async add({
+    profileId,
+    publicationId,
+    reactionType,
+  }: ReactionRequest): PromiseResult<void, NetworkError> {
+    try {
+      const { data } = await this.apolloClient.mutate<
+        AddReactionMutation,
+        AddReactionMutationVariables
+      >({
+        mutation: AddReactionDocument,
+        variables: {
+          publicationId,
+          profileId,
+          reaction: reactionTypeMap[reactionType],
+        },
+      });
+
+      if (!data) {
+        return failure(
+          new NetworkError(
+            new Error(`Failed to add reaction to a publication with id "${publicationId}"`),
```

Comment:
@cesarenaldi what would be the best way to handle errors here? I see that `NetworkError` is deprecated, but as we are very specific with error types across the codebase, what would you suggest to return here?
Possible errors are:
- server/network connection
- add/remove reaction which already exists
- wrong variables
- ??

Reply:
With hindsight `NetworkError` is not a domain failure mode.
It's a generic wrapper for unexpected errors that might happen during any operation.

The decision we made when adopting a code style that make use of `Result<T, E>` (which is getting widely adopted as primitive in many languages like Kotlin with his parametrized enums, Swift with it's equivalent of `Result`, but also Rust) is to make a clear distinction between "known failure modes" (i.e. normal way of operating that leads to a not successful result) and actual unexpected failures (which we usually refer to as traditional exceptions).

Known failure modes are piped back into the program's control flow via the `Result<T, E>` primitive (`Failure<T, E>` to be accurate). These becomes explicit and forces consumers of a given function/method to be explicit in the way these are handled.

Exceptions are instead thrown and bubbles up the call stack "as usual".

When using a 3rd party lib or a system API it's not uncommon to have to catch low level exception which are thrown, detect if a failure mode has been encountered and return a `Failure<T, E>` with a domain modelled error type.

To me, Server/Network connection errors for the class of problems we are solving falls into the "exception categories". Throw them, let them bubble up the call stack. In a React app the  call stack might lead to a render cycle method and they can be handled with a traditional React Error Boundary (*)

Add/remove reaction if can be programmatically cause in a normal way of operating seems to fall into known failure modes (so `failure` seems appropriate). If instead they can only be caused by a programming mistake I would consider a cheaper invariant error.

Wrong variables without more context seems to fall into a coding mistake... can it be done in a normal operating situation?

I guess IMO there is no one size fits all. It depends on the problem at hand and what are the actual failure modes (if there are any).

(*) we should/could consider to add an Error Boundary in the `LensProvider` at some point as catch-all in case the developer didn't manage they own application error boundaries.

Reply:
thank you for the explanation. I found out that in the case of `useReaction` hook, we have a presenter that has to know about possible failures of the API call to revert the optimistic reaction. that forces us to return a possible error in the `error` field. or shall we throw from the hook if an error is returned from the gateway?

right now it looks like here, but according to what you wrote, that is a coding mistake to allow to submit the same reaction multiple times.
![Screenshot 2023-01-02 at 15 37 58](https://user-images.githubusercontent.com/605420/210246240-d6d7c72d-dda6-4340-9563-7b89e9b3dd95.png)